### DISCUSSION TITLE
Android中弹框中保存数据问题

### DISCUSSION THREAD

**Discussion by 15821217725** [Tencent/MMKV#948] [Q&A] [answered]
1.第一次在dialog中保存数据
2.关掉app
3.第二次在oncreate中取不到
说明：第一次和第二次不要同时执行

**Comment by lingol** [comment, marked as answer]
Check the logs from logcat.

**Reply by 15821217725** [reply]
![image](https://user-images.githubusercontent.com/30583585/188530012-47b86878-8f3d-4e96-b49f-08737c8f806b.png)

**Reply by 15821217725** [reply]
这条数据是有的，但是取出是0，如果延迟能取到，如果在oncreate中保存能立即取到，在弹框中保存就不行，同进程同线程我想知道为什么

**Reply by lingol** [reply]
What do you mean by saying `dialog`, did it happen before or after your `Application.onCreate()`? All MMKV methods must be called after you have initialized MMKV in `Application.onCreate()`.

**Reply by lingol** [reply]
The log you quoted suggests that you have not written anything to the MMKV file `user_change_center_module`. Did you see any log about it when you store any key-value in the so-called `dialog`?

**Reply by 15821217725** [reply]
都是在Application.onCreate()之后发生的，我在dialog中有一个按钮，点击按钮的回调中写入了数据，下次冷启动在oncreate()中取不到写入的数据，您要说没写进去的话为什么延时1s后能取到数据呢。另外写入数据时同样也是报的“with 0 actual size, file size 4096, InterProcess 0, meta info version:0”这个日志。

**Reply by lingol** [reply]
Something very basic must be wrong in your code. Can you provide a demo project?

**Reply by lingol** [reply]
It's simply impossible for MMKV to fail to get value at one time but succeed after a 1s delay.

**Reply by 15821217725** [reply]
正常在onCreate中写入数据是能立即取到的，项目中业务很复杂，现在我排查出应该是弹框的问题，如果您需要，我待会单独写个demo给您演示。

**Reply by 15821217725** [reply]
我写了一个demo发现dialog中存入下次可以在oncreate中取到，但是在我的项目中确实取不到，如果不是在dialog中则么样都能取到。
排查发现在取数据后边又调用了MMKV.initialize(“path”)方法，并且传入了path，首次初始化是在application.oncreate中MMKV.initialize(context)，应该是这个影响到的，因为我注释了后边第二次init的方法就行了。

**Reply by lingol** [reply]
You should never initialize MMKV twice.

**Reply by 15821217725** [reply]
我知道，这是之前项目中的bug，像这样的情况初始化两次，应该会存在两个路径，弹框中点击回调写入时两个都已经初始化了，这时会写入到哪个目录下？如果只执行了application.oncreate中的初始化，先decode取一次，在进行第二次初始化，这时写入到哪个目录下。我发现如果每次在oncreate中取一次数据就不会有取不到的情况了，项目比较复杂，缓存中放了很多东西，改成一次初始化肯定会舍弃一个目录，这样我怕影响到之前的用户。

**Reply by 15821217725** [reply]
回复：It's simply impossible for MMKV to fail to get value at one time but succeed after a 1s delay.因为第二次初始化还没执行，1s后第二次初始化执行了就能取到了，如果保存数据在第二次初始化之前保存或者获取一次，就不用等待1s也能取到值了

**Reply by lingol** [reply]
>You should never initialize MMKV twice.

I won't repeat myself.
